### ISSUE TITLE
Tutorial: Nested resampling should probably be its own section in "advanced"

### ISSUE THREAD

**Issue by berndbischl** [mlr-org/mlr#233]

**Comment by rpinsler** [comment]
I have a question regarding nested resampling in a benchmark experiment (to justify the question here, maybe consider making this more clear in the tutorial). How do I properly set up an experiment that compares different algorithms which are tuned in the process and perform a prediction on a test set?

Clearly, I need an inner validation loop for parameter tuning. Do I need yet another outer validation when I compare the algorithms or can I directly provide a holdout/test set in the outer loop? How do I proceed after that? I feel there should be some function getBMRBestLearner according to a performance criterion.

I also think the process should not be so much different from what you would do without a benchmark. If I understand correctly, in a single learning task you call train() and predict() whereas in a benchmark experiment you call benchmark(). It would be more consistent if you first call train() and predict() for each task/learner and then combine the results into a benchmark (caret approach) or you call train()/benchmark() no matter if you have only one or multiple tasks/learners.

**Comment by schiffner** [comment]
Regarding the first half of your question: For predictions on a single test set, you can choose `"Holdout"` (`makeResampleDesc(method = "Holdout")`) for the outer resampling loop, or, as I guess you have a fixed test set, use `makeFixedHoldoutInstance` to provide the indices of train and test observations. The whole benchmark experiment (with only one learner and task) then would look like this:

```
#### tune k-nearest neighbor
## use cross-validation in inner loop
in.rdesc = makeResampleDesc("CV", iters = 3)
ps = makeParamSet(makeDiscreteParam("k", 3:5))
ctrl = makeTuneControlGrid()
lrn = makeTuneWrapper("classif.kknn", resampling = in.rdesc, par.set = ps, control = ctrl,
  show.info = FALSE)

## specify train and test data sets for outer loop
train.inds = c(1:30, 51:80, 101:130)
test.inds = setdiff(1:150, train.inds)
out.rin = makeFixedHoldoutInstance(train.inds = train.inds, test.inds = test.inds, size = 150)

## benchmark experiment
res = benchmark(lrn, iris.task, out.rin, measure = acc, show.info = FALSE)
res
```

Is this what you had in mind?

**Comment by rpinsler** [comment]
Yes, that's what I currently do. I guess I was a little bit confused because this common setup wasn't covered in the tutorial, so I thought I might do something wrong.

**Comment by schiffner** [comment]
I have started a Nested Resampling section in the tutorial. Currently, it contains

* nested resampling with hyperparameter tuning, 
* nested resampling with feature selection (wrapper approach),
* nested resampling with benchmarking with examples for tuning and feature selection.

I will mention prediction on a fixed test set and show an example. Is there anything else concerning benchmarking/nested resampling, what you feel is missing or should be covered more thoroughly in the tutorial?

**Comment by rpinsler** [comment]
The nested resampling tutorial itself seems alright. For benchmarking, I think if the process differs from the single task/single learner setup (see above), then this should be pointed out. Furthermore, it would be interesting to see an example of what you can do with the benchmark results, e.g. visualize performances and pick the best one (similar to: http://topepo.github.io/caret/training.html#resamp). However, it seems that currently there is also some visualization functionality missing or in progress which might fit here.

**Comment by schiffner** [comment]
Thank you very much for your comment and the link. I'm sorry for the long wait.

I have reworked the section about benchmark experiments (http://berndbischl.github.io/mlr/tutorial/html/benchmark_experiments/index.html).
It now covers prediction on a single test set (either randomly selected or fixed) and also shows how to extract and visualize performances.

We currently have someone working on more functionality for analysis of benchmark experiments, especially plots and statistical tests to rank learners and identify the best learner(s).

The section about nested resampling should be up by the end of the week.

**Comment by larskotthoff** [comment]
Did you have a chance to push this @schiffner ? Can we close this issue?

**Comment by berndbischl** [comment]
short ping
We could also move this to the tutorial tracker?

**Comment by berndbischl** [comment]
Although closing would be better :)

**Comment by larskotthoff** [comment]
Ok, closing.

**Comment by schiffner** [comment]
Finally!

https://github.com/mlr-org/mlr-tutorial/commit/f2531275dbd4fcc6a8cb0ad2bb23322e2ac8eb6f

**Comment by BabyBonda** [comment]
Hi @schiffner I need your help! I read your mrl tutorial and I have a problem with nested cross validation. I understand how can i do feature selection (wrapper method) with nested cross validation and tuning hyperparameters with nested cross validation in a separate way. 

What I want to do is a feature selection ( via wrapper method ) and tuning hyperparameters with a single nested cross validation ? Instead use two separate nested cross validation, one for feature selection and one for tuning, can I use one nested cross validation to do feature selection and tuning together?

I hope you understand my doubt.
Thank you very much!

Barbara.
